Update LDAP configuration docs (#13245)
druid-io/druid: docs/operations/auth-ldap.md

  ~ under the License.
  -->

You can use [Lightweight Directory Access Protocol (LDAP)](https://en.wikipedia.org/wiki/Lightweight_Directory_Access_Protocol) to secure access to Apache Druid. This topic describes how to set up Druid authentication and authorization with LDAP and LDAP over TLS (LDAPS). The examples on this page show the configuration for an Active Directory LDAP system.

The first step is to enable LDAP authentication and authorization for Druid. You then map an LDAP group to Druid roles and assign permissions to those roles. After you've completed this configuration you can optionally choose to enable LDAPS to make LDAP traffic confidential and secure.

## Prerequisites

Before you start to configure LDAP for Druid, test your LDAP connection and perform a sample search.

### Check your LDAP connection

> Verify that your LDAP connection is working by testing it with user credentials. For example, the following command tests the connection for the user `myuser@example.com`. Insert your LDAP server IP address and substitute `389` for the port number of your LDAP instance, if different.

Author: Updated.

```bash
ldapwhoami -vv -H ldap://ip_address:389  -D "myuser@example.com" -W
```

Enter the password for the user when prompted and verify that the command succeeded. If it failed, check the following:

- Make sure you're using the correct port for your LDAP instance.
- Check whether a network firewall is preventing connections to the LDAP port.
- Review your LDAP implementation details to see whether you need to specifically whitelist LDAP clients at the LDAP server. If so, add the Druid Coordinator server to the whitelist.

### Test your LDAP search

> Once your LDAP connection is working, search for a user. For example, the following command searches for the user myuser in an Active Directory system. The `sAMAccountName` attribute is specific to Active Directory and contains the authenticated user identity:

Author: @tijoparacka I don't understand this comment. Please suggest the correction.

> Once your LDAP connection is working, search for a user. For example, the following command searches for the user myuser in an Active Directory system. The `sAMAccountName` attribute is specific to Active Directory and contains the authenticated user identity:

Author: Updated.

```bash
ldapsearch -x -W -H ldap://ip_address:389  -D "cn=admin,dc=example,dc=com" -b "dc=example,dc=com" "(sAMAccountName=myuser)" +
```

> The `memberOf` attribute in the results shows the groups the user belongs to. For example, the following response shows that the user is a member of the `group1` group:

Author: @tijoparacka Can you please provide the partial output for me to include?

> The `memberOf` attribute in the results shows the groups the user belongs to. For example, the following response shows that the user is a member of the `group1` group:

Author: Updated.

```bash
memberOf: cn=group1,ou=groups,dc=example,dc=com
```

You use this information to map the LDAP group to Druid roles in a later step. 

> Druid uses the `memberOf` attribute to determine a group's membership using LDAP. If your LDAP server implementation doesn't include this attribute, you must complete some additional steps when you [map LDAP groups to Druid roles]().

## Configure Druid for LDAP authentication

To configure Druid to use LDAP authentication, follow these steps. See [Configuration reference](../configuration/index.md) for the location of the configuration files. 

> 1. Create a user in your LDAP system that you'll use for internal communication with Druid. See [Security overview](./security-overview.md) for more information.

Author: Updated.

   druid.auth.authenticator.ldap.credentialsValidator.type=ldap
   druid.auth.authenticator.ldap.credentialsValidator.url=ldap://ip_address:port
   druid.auth.authenticator.ldap.credentialsValidator.bindUser=administrator@example.com
   druid.auth.authenticator.ldap.credentialsValidator.bindPassword=adminpassword
   druid.auth.authenticator.ldap.credentialsValidator.baseDn=dc=example,dc=com
   druid.auth.authenticator.ldap.credentialsValidator.userSearch=(&(sAMAccountName=%s)(objectClass=user))
   druid.auth.authenticator.ldap.credentialsValidator.userAttribute=sAMAccountName
   druid.auth.authenticator.ldap.authorizerName=ldapauth
   druid.escalator.type=basic
   druid.escalator.internalClientUsername=internal@example.com
   druid.escalator.internalClientPassword=internaluserpassword
   druid.escalator.authorizerName=ldapauth
   druid.auth.authorizers=["ldapauth"]
   druid.auth.authorizer.ldapauth.type=basic
   druid.auth.authorizer.ldapauth.initialAdminUser=internal@example.com
   druid.auth.authorizer.ldapauth.initialAdminRole=admin
   druid.auth.authorizer.ldapauth.roleProvider.type=ldap
   ```
   Note the following:

>    - `bindUser`: A user for connecting to LDAP.

Author: Updated.

   - `userSearch`: Your LDAP search syntax.
   - `userAttribute`: The user search attribute.
   - `internal@example.com` is the LDAP user you created in step 1. In the example it serves as both the internal client user and the initial admin user.

   > In the above example, the [Druid escalator](../development/extensions-core/druid-basic-security.md#escalator) and LDAP initial admin user are set to the same user - internal@example.com. If the escalator is set to a different user, you must follow steps 4 and 5 to create the group mapping and allocate initial roles before the rest of the cluster can function.

4. Save your group mapping to a JSON file. An example file `groupmap.json` looks like this:
   
   ```
   {
      "name": "group1map",

>       "groupPattern": "CN=group1,CN=Users,DC=example,DC=com",

Author: Updated.

      "roles": [
         "readRole"
      ]
   }
   ```

>    In the example, the LDAP group `group1` maps to Druid role `readRole` and the name of the mapping is `group1map`.

Author: Updated.

> 5. Use the Druid API to create the group mapping and allocate initial roles according to your JSON file. The following example uses curl to create the mapping defined in `groupmap.json` for the LDAP group `group1`:

Author: Updated.

   
   ```

>    curl -i -v  -H "Content-Type: application/json" -u internal -X POST -d @groupmap.json http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/groupMappings/group1map

Author: Updated.

   ```
6. Check that the group mapping was created successfully. The following example request lists all group mappings:

   ```
   curl -i -v  -H "Content-Type: application/json" -u internal -X GET  http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/groupMappings
   ```

## Map LDAP groups to Druid roles

Once you've completed the initial setup and mapping, you can map more LDAP groups to Druid roles. Members of an LDAP group get access to the permissions of the corresponding Druid role.

> ### Step 1: Create a Druid role

Author: Updated.

> To create a Druid role, you submit a POST request to the Coordinator process using the Druid REST API.

Author: Updated.

Once you have a Druid role you can add permissions to it. The following example adds read-only access to a `wikipedia` data source.

Given the following JSON in a file named `perm.json`:

```
[
	{ "resource": { "name": "wikipedia", "type": "DATASOURCE" }, "action": "READ" },
    { "resource": { "name": ".*", "type": "STATE" }, "action": "READ" },
	{ "resource": {"name": ".*", "type": "CONFIG"}, "action": "READ"}
]
```

The following request associates the permissions in the JSON file with the `readRole` role:

```
curl -i -v  -H "Content-Type: application/json" -u internal -X POST -d@perm.json  http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/roles/readRole/permissions
```

Druid users need the `STATE` and `CONFIG` permissions to view the data source in the Druid console. If you only want to assign querying permissions you can apply just the `READ` permission with the first line in the `perm.json` file.

You can also provide the data source name in the form of a regular expression. For example, to give access to all data sources starting with `wiki`, you would specify the data source name as `{ "name": "wiki.*" }` .

### Step 3: Create the group mapping

> You can now map an LDAP group to the Druid role. The following example request creates a mapping with name `group1map`. It assumes that a group named `group1` exists in the directory.

Author: Updated.

```
{

>     "name": "group1map",

Author: Updated.

>     "groupPattern": "CN=group1,CN=Users,DC=example,DC=com",

Author: Updated.

    "roles": [
        "readRole"
    ]
}
```

The following example request configures the mapping&mdash;the role mapping is in the file `groupmap.json`. See [Configure Druid for LDAP authentication](#configure-druid-for-ldap-authentication) for the contents of an example file.

```

> curl -i -v  -H "Content-Type: application/json" -u internal -X POST -d @groupmap.json http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/groupMappings/group1map

Author: Updated.

```

To check whether the group mapping was created successfully, the following request lists all group mappings:

```
curl -i -v  -H "Content-Type: application/json" -u internal -X GET http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/groupMappings
```

The following example request returns the details of the `group1map` group:

```

> curl -i -v  -H "Content-Type: application/json" -u internal -X GET http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/groupMappings/group1map

Author: Updated.

```

> The following example request adds the role `queryRole` to the `group1map` mapping:

Author: Updated.

```

> curl -i -v  -H "Content-Type: application/json" -u internal -X POST http://localhost:8081/druid-ext/basic-security/authorization/db/ldapauth/groupMappings/group1/roles/queryrole

Author: Updated.

```

> ### Step 4: Add an LDAP user to Druid and assign a role

Reviewer: Line 216 to 231 is to map a user to role.  This section of the document is about how to map LDAP group with  druid roles hence  Line 216 to 231 is not required.

Author: @tijoparacka the original document for this step included the proviso - can you please confirm that this section is still not needed? What do you do if the points below apply?

You only need to complete this step if:
- Your LDAP server doesn't support the `memberOf` attribute, or
- You want to configure a user with additional Druid roles that are not mapped to the LDAP groups that the user belongs to.

Author: Updated after discussion with Tijo.

You only need to complete this step if:
- Your LDAP server doesn't support the `memberOf` attribute, or
- You want to configure a user with additional Druid roles that are not mapped to the LDAP groups that the user belongs to.

Example request to add the LDAP user `myuser` to Druid:

```
curl -i -v  -H "Content-Type: application/json" -u internal -X POST http://localhost:8081/druid-ext/basic-security/authentication/db/ldap/users/myuser 
```

Example request to assign the `myuser` user to the `queryRole` role:

```
curl -i -v  -H "Content-Type: application/json" -u internal -X POST http://localhost:8081/druid-ext/basic-security/authentication/db/ldap/users/myuser/roles/queryRole
```

## Enable LDAP over TLS (LDAPS)

Once you've configured LDAP authentication in Druid, you can optionally make LDAP traffic confidential and secure by using Transport Layer Security (TLS)&mdash;previously Secure Socket Layer(SSL)&mdash;technology. 

## Prerequisites

Before you start to set up LDAPS in Druid:

- You must [configure Druid for LDAP authentication](#configure-druid-for-ldap-authentication).

> - You need a CA certificate for your LDAP server.

Reviewer: Either a second or third step is required. ie either a CA certificate for the LDAP server or if the LDAP is configured with Self-signed certificate then the same certificate needs to be used in druid truststore.

Author: Updated.

> - You need a CA certificate for your LDAP server.

Reviewer 2: I get a little lost in the certificate section.

We are trying to establish trust between Druid and the LDAP server.

The LDAP server should be secured with a certificate issued by a public CA (certificate authority) or self-signed/certified by an internal CA (which is common case for internal-only services like LDAP)

You have to  import into the Druid trust store:
a) The LDAP server public certificate (regardless of who signed it)
b) the root certificate for the the CA that signed the certificate for the LDAP server. If it is a common public CA, this certificate may already be in the java truststore. Otherwise you need to import the certificate for the CA.

If you import the Root CA certificate, you're indicating you trust any certificate signed by that authority. When you import the LDAP server certificate, you're indicating that you trust the LDAP server based upon the authority of the CA

Author: Updated this section.

- You need a self-signed certificate, or an Active Directory(AD) certificate if you configured LDAP using AD.

## Configure Druid for LDAPS

Complete the following steps to set up LDAP for Druid. See [Configuration reference](../configuration/index.md) for the location of the configuration files. 

1. Import the CA certificate for your LDAP server into the location saved as your `druid.client.https.trustStorePath` in your `common.runtime.properties` file.

   ```
   keytool -import -trustcacerts -keystore path/to/cacerts -storepass truststorepassword -alias aliasName -file path/to/certificate.cer
   ```

>    Replace `path/to/cacerts` with the path to your trust store, `truststorepassword` with your trust store password, `aliasName` with an alias name for the keystore, and `path/to/certificate.cer` with the location and name of your certificate. For example:

Reviewer: truststore is a single word

Author: Updated.

   ```
   keytool -import -trustcacerts -keystore /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/jre/lib/security/cacerts -storepass mypassword -alias myAlias -file /etc/ssl/certs/my-certificate.cer
   ```

2. Import your self-signed certificate or Active Directory certificate:

   ```
   keytool -importcert -keystore path/to/cacerts -storepass truststorepassword -alias aliasName -file path/to/certificate.cer
   ```

>    Replace `path/to/cacerts` with the path to your trust store, `truststorepassword` with your trust store password, `aliasName` with an alias name for the keystore, and `path/to/certificate.cer` with the location and name of your certificate. For example:

Reviewer: truststore to be a single word

Author: Updated.

  ~ under the License.
  -->

You can use [Lightweight Directory Access Protocol (LDAP)](https://en.wikipedia.org/wiki/Lightweight_Directory_Access_Protocol) to secure access to Apache Druid. This topic describes how to set up Druid authentication and authorization with LDAP and LDAP over TLS (LDAPS). The examples on this page show the configuration for an Active Directory LDAP system.

The first step is to enable LDAP authentication and authorization for Druid. You then map an LDAP group to Druid roles and assign permissions to those roles. After you've completed this configuration you can optionally choose to enable LDAPS to make LDAP traffic confidential and secure.

## Prerequisites

Before you start to configure LDAP for Druid, test your LDAP connection and perform a sample search.

### Check your LDAP connection

> Verify that your LDAP connection is working by testing it with user credentials. Later in the process you [configure Druid for LDAP authentication](#configure-druid-for-ldap-authentication) with this user as the `bindUser`.

Reviewer 2: ```suggestion
Test your LDAP connection to verify it works with user credentials. Later in the process you [configure Druid for LDAP authentication](#configure-druid-for-ldap-authentication) with this user as the `bindUser`.
```

The following example command tests the connection for the user `myuser@example.com`. Insert your LDAP server IP address. Modify the port number of your LDAP instance if it listens on a port other than `389`.

```bash
ldapwhoami -vv -H ldap://ip_address:389  -D "myuser@example.com" -W
```

Enter the password for the user when prompted and verify that the command succeeded. If it failed, check the following:

- Make sure you're using the correct port for your LDAP instance.

> - Check whether a network firewall is preventing connections to the LDAP port.

Reviewer 2: ```suggestion
- Check if a network firewall is preventing connections to the LDAP port.
```

> - Review your LDAP implementation details to see whether you need to specifically whitelist LDAP clients at the LDAP server. If so, add the Druid Coordinator server to the whitelist.

Reviewer 2: ```suggestion
- Review your LDAP implementation details to see whether you need to specifically allow LDAP clients at the LDAP server. If so, add the Druid Coordinator server to the allow list.
```
avoid "whitelist", "blacklist"
